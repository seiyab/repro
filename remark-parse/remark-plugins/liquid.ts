import { CompileContext, Token } from "mdast-util-from-markdown";
import { codes, types } from "micromark-util-symbol";
import { Code, Effects, State } from "micromark-util-types";
import { Processor } from "unified";

function remarkLiquid(this: Processor): void {
	const data = this.data();

	(data.micromarkExtensions ??= []).push(syntax());
	(data.fromMarkdownExtensions ??= []).push(fromMarkdown());
}

function syntax(): any {
	return {
		text: {
			[codes.leftCurlyBrace]: {
				name: "liquid",
				tokenize: liquidTokenize,
			},
		},
	};

	function liquidTokenize(effects: Effects, ok: State, nok: State) {
		return start;

		function start(code: Code) {
			effects.attempt;
			effects.enter("liquid" as any);
			effects.enter(types.data);
			effects.consume(code);
			return function (code: Code) {
				if (code === codes.rightCurlyBrace) {
					return nok;
				}
				return inside;
			};
		}

		function inside(code: Code) {
			if (code === codes.rightCurlyBrace) {
				effects.consume(code);
				effects.exit(types.data);
				effects.exit("liquid" as any);
				return ok;
			}
			if (code === codes.eof) {
				return nok;
			}
			effects.consume(code);
			return inside;
		}
	}
}

function fromMarkdown() {
	/** @type {Extension} */
	return {
		canContainEols: ["liquid"],
		enter: { liquid: enterLiquid },
		exit: { liquid: exitLiquid },
	};

	function enterLiquid(this: CompileContext, token: Token) {
		this.enter({ type: "liquid" } as any, token);
		this.buffer();
	}

	function exitLiquid(this: CompileContext, token: Token) {
		const d = this.resume();
		const node = this.stack[this.stack.length - 1];
		node.value = d;
		this.exit(token);
	}
}

export { remarkLiquid };
