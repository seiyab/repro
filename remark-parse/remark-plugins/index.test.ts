import { describe, test, expect } from "bun:test";
import remarkParse from "remark-parse";
import { unified } from "unified";
import { remarkLiquid } from "./liquid";

describe("liquid", () => {
	function parse(text: string) {
		const processor = unified().use(remarkParse).use(remarkLiquid);

		return processor.run(processor.parse(text));
	}

	test.each([
		"abc{def}geh",
		"aa{[link-like](example.com)}bb",
		"a\nb",
		"{",
		"{}",
		`a{
    b
    c
  }d`,
	])("%s", async (t) => {
		expect(await parse(t)).toMatchSnapshot();
	});
});
