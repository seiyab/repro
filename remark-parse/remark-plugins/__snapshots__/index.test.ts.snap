// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`liquid abc{def}geh 1`] = `
{
  "children": [
    {
      "children": [
        {
          "position": {
            "end": {
              "column": 4,
              "line": 1,
              "offset": 3,
            },
            "start": {
              "column": 1,
              "line": 1,
              "offset": 0,
            },
          },
          "type": "text",
          "value": "abc",
        },
        {
          "position": {
            "end": {
              "column": 9,
              "line": 1,
              "offset": 8,
            },
            "start": {
              "column": 4,
              "line": 1,
              "offset": 3,
            },
          },
          "type": "liquid",
          "value": "{def}",
        },
        {
          "position": {
            "end": {
              "column": 12,
              "line": 1,
              "offset": 11,
            },
            "start": {
              "column": 9,
              "line": 1,
              "offset": 8,
            },
          },
          "type": "text",
          "value": "geh",
        },
      ],
      "position": {
        "end": {
          "column": 12,
          "line": 1,
          "offset": 11,
        },
        "start": {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "paragraph",
    },
  ],
  "position": {
    "end": {
      "column": 12,
      "line": 1,
      "offset": 11,
    },
    "start": {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`liquid aa{[link-like](example.com)}bb 1`] = `
{
  "children": [
    {
      "children": [
        {
          "position": {
            "end": {
              "column": 3,
              "line": 1,
              "offset": 2,
            },
            "start": {
              "column": 1,
              "line": 1,
              "offset": 0,
            },
          },
          "type": "text",
          "value": "aa",
        },
        {
          "position": {
            "end": {
              "column": 29,
              "line": 1,
              "offset": 28,
            },
            "start": {
              "column": 3,
              "line": 1,
              "offset": 2,
            },
          },
          "type": "liquid",
          "value": "{[link-like](example.com)}",
        },
        {
          "position": {
            "end": {
              "column": 31,
              "line": 1,
              "offset": 30,
            },
            "start": {
              "column": 29,
              "line": 1,
              "offset": 28,
            },
          },
          "type": "text",
          "value": "bb",
        },
      ],
      "position": {
        "end": {
          "column": 31,
          "line": 1,
          "offset": 30,
        },
        "start": {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "paragraph",
    },
  ],
  "position": {
    "end": {
      "column": 31,
      "line": 1,
      "offset": 30,
    },
    "start": {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`liquid a
b 1`] = `
{
  "children": [
    {
      "children": [
        {
          "position": {
            "end": {
              "column": 2,
              "line": 2,
              "offset": 3,
            },
            "start": {
              "column": 1,
              "line": 1,
              "offset": 0,
            },
          },
          "type": "text",
          "value": 
"a
b"
,
        },
      ],
      "position": {
        "end": {
          "column": 2,
          "line": 2,
          "offset": 3,
        },
        "start": {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "paragraph",
    },
  ],
  "position": {
    "end": {
      "column": 2,
      "line": 2,
      "offset": 3,
    },
    "start": {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`liquid { 1`] = `
{
  "children": [
    {
      "children": [
        {
          "position": {
            "end": {
              "column": 2,
              "line": 1,
              "offset": 1,
            },
            "start": {
              "column": 1,
              "line": 1,
              "offset": 0,
            },
          },
          "type": "text",
          "value": "{",
        },
      ],
      "position": {
        "end": {
          "column": 2,
          "line": 1,
          "offset": 1,
        },
        "start": {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "paragraph",
    },
  ],
  "position": {
    "end": {
      "column": 2,
      "line": 1,
      "offset": 1,
    },
    "start": {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`liquid {} 1`] = `
{
  "children": [
    {
      "children": [
        {
          "position": {
            "end": {
              "column": 3,
              "line": 1,
              "offset": 2,
            },
            "start": {
              "column": 1,
              "line": 1,
              "offset": 0,
            },
          },
          "type": "text",
          "value": "{}",
        },
      ],
      "position": {
        "end": {
          "column": 3,
          "line": 1,
          "offset": 2,
        },
        "start": {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "paragraph",
    },
  ],
  "position": {
    "end": {
      "column": 3,
      "line": 1,
      "offset": 2,
    },
    "start": {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`liquid a{
    b
    c
  }d 1`] = `
{
  "children": [
    {
      "children": [
        {
          "position": {
            "end": {
              "column": 2,
              "line": 1,
              "offset": 1,
            },
            "start": {
              "column": 1,
              "line": 1,
              "offset": 0,
            },
          },
          "type": "text",
          "value": "a",
        },
        {
          "position": {
            "end": {
              "column": 4,
              "line": 4,
              "offset": 18,
            },
            "start": {
              "column": 2,
              "line": 1,
              "offset": 1,
            },
          },
          "type": "liquid",
          "value": 
"{
    b
    c
  }d"
,
        },
      ],
      "position": {
        "end": {
          "column": 5,
          "line": 4,
          "offset": 19,
        },
        "start": {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "paragraph",
    },
  ],
  "position": {
    "end": {
      "column": 5,
      "line": 4,
      "offset": 19,
    },
    "start": {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;
